teps/0027-task-execution-status-at-runtime.md

```diff
@@ -0,0 +1,133 @@
+
+An additional but important goal of this proposal is identify the different execution states of a `pipelineTask` and
+how each of those states derived based on `taskRun` and/or `pipelineRun` status.
+
+### Non-Goals
+
+* This proposal is not to expose the execution state of `finally` tasks since its not required.
+ 
+* This proposal is not to expose the entire status of a `taskRun`/`pipelineRun` as metadata at runtime.
+
+## Proposal
+
+Introduce a `param` value `$(context.pipelineRun.Tasks.<pipelineTask>)` which resolves to one of the execution states:
+`Succeeded`, `Failed`, `Skipped`, and `None`
+
+This `param` value is populated and available at the runtime. In the following example, `deployment` is a `dag` task and `rollback` is a `finally` task. The `finally` task, `rollback`, checks the execution status of `deployment` task using context `param` and continues if `deployment` resulted in failure. 
+
+```yaml
+    finally:
+    - name: rollback
+      params:
+        - name: deployment-state
+          value: "$(context.pipelineRun.Tasks.deployment)"
+      taskSpec:
+        params:
+          - name: deployment-state
+        steps:
+          - image: ubuntu
+            name: check-deployment
+            script: |
+              if [ $(params.deployment-state) == "Failed" ]
+              then
+                echo "Deployment has failed, it's time to rollback!"
+              fi
+```
+
+The same `param` will be made available in the `dag` tasks at runtime. There might be less known use cases in which
+`dag` tasks need access to the execution status of other `dag` tasks. In general, `dag` tasks are designed using
+`runAfter`, `from`, and `task results` to build chain of dependencies, and the execution continues only if all of them
+are successful. One single failure triggers `finally` tasks which needs to be aware of such failure of `dag` task.
+
+The `dag` tasks can also be designed as guarded tasks using `when expressions`. Such guarded tasks are `skipped` if
+`when expressions` evaluate to false but `pipeline` continues executing. There might be scenarios where a `dag` task
+would like to know whether the guraded task was `skipped` or `successful`.
+
+| State | Description |
+| ----- | ----------- |
+| `Succeeded` | The `pipelineTask` was successful i.e. a respective pod was created and completed successfully. The `pipelineTask` had a `taskRun` with `Succeeded` `ConditionType` and  `True` `ConditionStatus`. |
+| `Failed` | The `pipelineTask` failed i.e. a respective pod was created but exited with error. The `pipelineTask` has a `taskRun` with `Succeeded` `ConditionType`,  `False` `ConditionStatus` and have exhausted all the retries. |
+| `Skipped` | The `pipelineTask` is `skipped` because of `when expression` or one of the parent tasks was `skipped`. It is part of `pipelineRun.Status.SkippedTasks`. |
+| `None` | This state means `none of the above`. |
```

Comment:
Now, since the scope of this TEP is only limited to `finally` tasks having access to execution state of `dag` tasks, `Running` state is no longer applicable. So far we have four states:

* Succeeded
* Failed
* Skipped
* None (not executed)

If we drop `Skipped` out of this list:

* Succeeded
* Failed
* None (not executed or skipped)

Reply:
I have narrowed it down to three states for the first iteration, can add more states as we go along:

```
Succeeded
Failed
None (not executed or skipped)
```